packages/toolpad-app/src/appDom/index.ts

```diff
@@ -727,14 +727,34 @@ export function moveNode<Parent extends AppDomNode, Child extends AppDomNode>(
   return setNodeParent(dom, node, parent.id, parentProp, parentIndex);
 }
 
-export function duplicateNode<Child extends ElementNode>(dom: AppDom, node: Child) {
+export function duplicateNode<Parent extends AppDomNode, Child extends ElementNode>(
+  dom: AppDom,
+  node: Child,
+  parent?: Parent,
+) {
   if (!node.parentId || !node.parentProp || !node.parentIndex) {
     throw new Error(`Node: "${node.id}" can't be duplicated`);
   }
 
+  const { children } = getChildNodes(dom, node);
+
   const newNode = createElement(dom, node.attributes.component!.value, node.props, node.layout);
 
-  return setNodeParent(dom, newNode, node.parentId, node.parentProp, node.parentIndex);
+  let updatedDom = dom;
+
+  if (Array.isArray(children)) {
+    children.forEach((childNode) => {
+      updatedDom = duplicateNode(updatedDom, childNode, newNode);
+    });
+  }
+
+  if (parent) {
```

Comment:
The naming might be confusing - parent index is actually a property of the child. It defines the position of that child inside its parent. So you could probably just use `addNode` instead of `seNodeParent` like Jan suggested.

Reply:
> Why do we have this check that still uses the same parent index?

Because when I duplicate children `parentIndex` remains the same, but `parent` node is a newly created node which is passed only when we duplicate children nodes

> I've noticed that if I duplicate an element inside a column for example, it's still being placed before/above the copied element, not after it, because it has the same parent index.

Can you clarify what you mean, because when I duplicate element it's placed after as can be seen in the recording attached. Maybe you are expecting that duplicated node would be selected afterwards? 🤔 Though technically I don't see a big difference as in this current flow I'd expect you'd still have to move new node into correct place

https://user-images.githubusercontent.com/437214/196437263-836b8a9c-f539-44d2-9fbb-8f8402571403.mov

Reply:
I see - but even duplicated components without children can have a parent, right? And for those we would want to set a new parent index...
I guess we might need to pass in a `parentIndex` to the `duplicateNode` function in order to be able to set if it should be a new index or not, according to each scenario? Unless you can think of a better solution, but this one sounds good to me at first glance.

About the video - I think that if a new component is added with the same parent index as another, if the components are inside a row it shows up to the right of the previous component, but inside a column it shows up above it.

https://user-images.githubusercontent.com/10789765/196443341-9a75837c-13d4-43d8-a5f2-000482961ebb.mov

Reply:
> Can you clarify what you mean, because when I duplicate element it's placed after as can be seen in the recording attached. Maybe you are expecting that duplicated node would be selected afterwards? 🤔 Though technically I don't see a big difference as in this current flow I'd expect you'd still have to move new node into correct place

I see what you mean here now too, yeah i thought the selected component was the new one.
Anyway I think this behavior is arbitrary even if it works visually... We should probably set a different parent index where possible.

Reply:
> (maybe it should be?)

I'm not really sure, I think it doesn't make much difference since it's identical 🤔 

> I guess we might need to pass in a parentIndex to the duplicateNode function in order to be able to set if it should be a new index or not, according to each scenario?

Well the way I understand how currently code works, we don't need to change `parentIndex` when we just duplicate children of that node, I mean those duplicated children has to have same index as the original, that's why it's using same.

> I think that if a new component is added with the same parent index as another, if the components are inside a row it shows up to the left of the previous component, but inside a column it shows up above it. Anyway this behavior is not intentional and we should not have components with the same parent index inside the same parent.

When a component (not children) is duplicated, it's `parentIndex` is set to new, so they aren't the same. As I explained above, only when we are duplicating children their `parentIndex` remains the same, because we aren't changing their order.

I may not understand if there is an issue right now with this code, it would be helpful if you could maybe then pull the local version to point exactly what might be a problem. As of now I think it's all working fine 🤔

Reply:
Actually it might be fine as it is, I guess I was just expecting the new component to be the selected one.